README.md
# MUiBalloons-website

#includes

    1. iBalloons Website (Static)
    2. myBalloons System (Web application)

### live now at: https://

## About

MUiBalloons-website source code
MUiBalloons-myBalloonsSystem source code
Developed using Django Framework

### This project is licensed under the terms of the Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International Public License (CC BY-NC-SA 4.0).

Please check the LICENSE.md for details.

### Development Status: ~~**Suspended until December 17, 2022 [*Expected to be back on progress on December 15, 2022*]**~~ Developing

### For development history please check changelog.md

Current Status: **Initial Development#7**
Current Version: **0.2**
Updated: **January 6, 2023**

---

## Instructions

#clone the repo
`git clone`

#create vitual environment
`python -m venv venv`

#activate vitual environment (macOS)
`source venv/bin/activate`

#install project dependencies
`pip install -r requirements.txt`

#run!
`python main.py`

---


changelog.md
# MU iBalloons, iBalloons web and myBalloons System update history

## The project is developed by Kittipich "Luke" Aiumbhornsin

## Created on November 19, 2022

### Config the System version in **init**.py

- Initial Create: initial created on November 19, 2022 -> 0.01

- Initial Development: Root-view template draft, custom error landing pages implemented on November 19, 2022 -> 0.1

- Initial Development#1.2: Custom error landing pages updated on November 20, 2022 -> 0.11

- Initial Development#1.3: Navigation bar and footer updated on November 20, 2022 -> 0.12

- Initial Development#2: Error page background image added and footer updated on November 21, 2022 -> 0.125

- Initial Development#2.2: Remove the secret data in **init**.py as the source code exit in public and reset the git repo to prevent the sensitive data shown in git repo on November 21, 2022 -> 0.13

- Initial Development#2.3: Home page view added on November 21, 2022 -> 0.132

- Initial Development#2.4: Dark Theme available for Home page on November 21, 2022 -> 0.133

- Initial Development#3: Login and Sign-Up page draft added on November 22, 2022 -> 0.135

- Initial Development#3.1: Add links to buttons in home page and contact button added to each error landing pages on November 22, 2022 -> 0.1351

- Initial Development#3.2: Update navbar, sign-up form, and add more database tables on November 22, 2022 -> 0.1352

- Initial Development#4: Change the app config (secret_key) with the value of env var on November 23, 2022 -> 0.1353

- Initial Development#4.2: User Role draft implemented, other bugs fixed, and Login System(Sign-Up, Login, Logout) works on November 23, 2022 -> 0.136

- Initial Beta#1: Complete Admin-User-Guest Permission restrictions, and overall improvements on November 24, 2022 -> 0.138

- Initial Fixed#1: Fix bugs at sign-up where master key is not accepting None type, iTicketViewer added, and others improvements on November 24, 2022 -> 0.14

- Initial Development#5: Sign-up form view customized, iTicketViewer interface added, and added brief login credential on navbar and navbar menu on Novenber 24, 2022 -> 0.142

- Initial Development#5.1: Update BootStrap version from **"5.2.2"** to **"5.2.3"**, make a draft of the page and add page restriction on Admin Dashboard, other improvements on November 25, 2022 -> 0.143

- Initial Development#5.2: User&Admin Dashboard Layout draft, database table changes, other improvements on November 25, 2022 -> 0.144

- Initial Development#6: Overall improvements on November 26, 2022 -> 0.1442

~~## Development has been suspended until December 17, 2022 [EXPECT to back on progress on December 15, 2022]~~

(added PyScript CDN on December 14, 2022)

- Initial Development#7: Preparing for develop using Django and preparing for new blueprint and plan on January 6, 2023 -> 0.2

TO-DO:

- [x] Implement custom navigation bar
- [x] Login and Signup pages
- [ ] Home page (Partially [DONE] -> shown texts needed to revised if needed)
- [ ] Admin&User Dashboard [Expect to start at v0.142]
- [ ] iTicketViewer
- [ ] Add page restriction to all that needed
- [ ] Notification System
- [ ] Profile (Avatar) System
- [ ] Login Session Timeout


outdated(flask)/myBalloons/__init__.py
# Root file of the system
# from flask_admin.contrib.sqla import ModelView
from flask_sqlalchemy import SQLAlchemy
db = SQLAlchemy()
# from flask_admin import Admin, BaseView, expose
# from flask_security import Security, SQLAlchemyUserDatastore
from flask_bcrypt import Bcrypt, generate_password_hash
from flask_login import LoginManager, current_user
from flask import Flask, Blueprint, render_template, abort, flash
from .models import User, TheKeys
from decouple import config as en_var # import the environment var
# from werkzeug.exceptions import HTTPException

DB_NAME = "myBalloons_database.sqlite"

def create_app():
    app = Flask(__name__)
    f_bcrypt = Bcrypt(app)
    app.config['FLASK_ADMIN-SWATCH'] = 'cerulean'
    app.config['SECRET_KEY'] = en_var('myBalloonsAppSecretKey') # Encrepted with Environment Variable
    app.config['DATABASE_NAME'] = DB_NAME
    app.config['SQLALCHEMY_DATABASE_URI'] = f'sqlite:///{DB_NAME}'
    app.config['SQLALCHEMY_TRACK_MODIFICATIONS'] = False
    app.config['TIMEZONE'] = 'Asia/Bangkok'
    """ # app.config['SECURITY_TRACKABLE'] = True
    # app.config['SECURITY_CONFIRMABLE'] = True
    # app.config['SECURITY_REGISTER_URL'] = '/sign-up/'
    # app.config['SECURITY_LOGIN_URL'] = '/login/'
    # app.config['SECURITY_LOGOUT_URL'] = '/logout/'
    # app.config['SECURITY_RESET_URL'] = '' """
    
    db.init_app(app)

    # from .account import account
    
    # from .master_tools import MTs

    """ # admin = Admin(app, name='MU iBalloons Database Manager', template_mode='bootstrap3')
    # admin.add_view(ModelView(User, db.session))
    # admin.add_view(ModelView(Login_log, db.session))
    # admin.add_view(ModelView(Ticket, db.session))
    # admin.add_view(ModelView(Note, db.session))
    # admin.add_view(ModelView(Participant, db.session))  """
    from .views.errorHandling import not_found, server_error, bad_request, forbidden, unauthorized, server_down
    app.register_error_handler(400, bad_request)
    app.register_error_handler(401, unauthorized)
    app.register_error_handler(403, forbidden)
    app.register_error_handler(404, not_found)
    app.register_error_handler(500, server_error)
    app.register_error_handler(503, server_down)

    from .views.general import views
    from .views.authen import auth
    from .views.dashboard_admin import admin_dashboard
    from .views.dashboard_user import user_dashboard
    app.register_blueprint(rootView, url_prefix='/')
    app.register_blueprint(views, url_prefix='/')
    app.register_blueprint(auth, url_prefix='/')
    app.register_blueprint(admin_dashboard, url_prefix='/')
    app.register_blueprint(user_dashboard, url_prefix='/')

    with app.app_context(): # Drop all of the tables
        db.drop_all()

    with app.app_context():
        db.create_all()


    @app.before_first_request
    def demo_account():
        try:

            k1 = TheKeys(key='master')
            k2 = TheKeys(key='admin')
            d1 = User(email="demo@admin.com", fname="Admin", lname="Admin Lastname", password=generate_password_hash("admin").decode('utf-8'),
             role_level=1, s_question='question', s_answer='answer')
            
            d2 = User(email='demo@user.com',fname="User", lname="User Lastname", 
            password=generate_password_hash("user").decode('utf-8'),s_question='question', 
            s_answer='answer')    
            
            db.session.add_all([k1,k2,d1, d2])
            db.session.commit()
        except Exception as e:
            flash(f'{e}', category='error')
            

    login_manager = LoginManager()
    login_manager.login_view = 'auth.logIn'
    login_manager.init_app(app)

    @login_manager.user_loader
    def load_user(id):
        return User.query.get(int(id))

    """ # # Setup Flask-Security
    # user_datastore = SQLAlchemyUserDatastore(db, User, Role)
    # security = Security(app, user_datastore)
    
    
    # @app.before_first_request
    # def create_user():
    #     db.create_all()
    #     user_datastore.create_user(email='matt@nobien.net', password='password')
    #     db.session.commit() """



    return app

class About():
    version = float()
    status = str()
    build = int()
    version_note = str()

    def __init__(self, version: float = float(0.0), status: str = 'None Stated', build: int = 20221100, version_note: str = "None Stated"):
        self.version = version
        self.status = status
        self.build = build
        self.version_note = version_note

    def __repr__(self) -> str:
        return str("{ " + f"Version: {self.version} | Status: {self.status} | Build: {self.build} | Details: {self.version_note}" + " }")

    def getSystemVersion(self) -> str:
        return str(self.version)

systemInfoObject = About(version=0.1442, status='Initial Development#6',
                         build=20221126, version_note='Overall improvements')
systemInfo = systemInfoObject.__repr__()
systemVersion = systemInfoObject.getSystemVersion()

rootView = Blueprint('rootView', __name__)
@rootView.route("/..root-template-view/")
def root_view():
    return render_template("base.html", about=systemInfo, user=current_user)

# Initial Development#6: Overall improvements on November 26, 2022 -> 0.1442